lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
 
-        // Only override parser if it is passed explicitly through the command line or if it's not
-        // defined yet (because the final object will at least have the parser key)
-        if (this.parser || !config.parser) {
-            config = ConfigOps.merge(config, {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
+                rootPath = path.dirname(localConfigFile);
+            }
 
-        if (this.parserOptions) {
-            config = ConfigOps.merge(config, {
-                parserOptions: this.parserOptions
-            });
+            debug(`Using ${localConfigFile}`);
+            configs.push(localConfig);
+            searched.push(localConfigDirectory);
         }
 
-        // Step 8: Merge in command line plugins
-        if (this.options.plugins) {
-            debug("Merging command line plugins");
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
 
-        // Step 9: Apply environments to the config if present
-        if (config.env) {
-            config = ConfigOps.applyEnvironments(config);
+            // Fall back on the personal config from ~/.eslintrc
+            debug("Using personal config file");
+            const personalConfig = this.getPersonalConfig();
+
+            if (personalConfig) {
+                configs.push(personalConfig);
+            } else if (!hasRules(this.options) && !this.options.baseConfig) {
+
+                // No config file, no manual configuration, and no rules, so error.
+                const noConfigError = new Error("No ESLint configuration found.");
+
+                noConfigError.messageTemplate = "no-config-found";
+                noConfigError.messageData = {
+                    directory,
+                    filesExamined: localConfigFiles
+                };
+
+                throw noConfigError;
+            }
         }
 
-        this.cache[directory] = config;
+        // Merged with any cached portion
+        configs.reverse();
+        cache = cache ? cache.concat(configs) : configs;
 
-        return config;
+        // Set the caches for the parent directories
+        configCache.setHierarchyLocalConfigs(searched, cache);
+
+        return cache;
     }
 
     /**
-     * Find local config files from directory and parent directories.
+     * Gets the vector of applicable configs and subconfigs from the hierarchy for a given file. A vector is an array of
+     * objects, each containing a config file path and an array of override indices corresponding to entries in config
+     * file's overrides section whose glob patterns match the specified file path; e.g., the vector
+     * { configFile: '/home/john/app/.eslintrc', matchingOverrides: [0, 2] } would indicate that the main project
```

Comment:
Is a vector an array of objects, or a single object? Based on the top of the comment it seems like it's an array of objects, but then this line is a bit misleading because it calls a single object a vector.